x/mongo/driver/operation.go

```diff
@@ -1260,6 +1250,35 @@ func (op Operation) addClusterTime(dst []byte, desc description.SelectedServer)
 	// return bsoncore.AppendDocumentElement(dst, "$clusterTime", clusterTime)
 }
 
+// calculateMaxTimeMS calculates the value of the 'maxTimeMS' field to potentially append
+// to the wire message based on the current context's deadline and the 90th percentile RTT
+// if the ctx is a Timeout context. If the context is not a Timeout context, it uses the
+// operation's MaxTimeMS if set. If no MaxTimeMS is set on the operation, and context is
+// not a Timeout context, calculateMaxTimeMS returns 0.
+func (op Operation) calculateMaxTimeMS(ctx context.Context, rtt90 time.Duration) (uint64, error) {
+	if internal.IsTimeoutContext(ctx) {
+		if deadline, ok := ctx.Deadline(); ok {
+			remainingTimeout := time.Until(deadline)
+			// We check that remainingTimeout <= rtt90 because in the rare case that they are equal, we
+			// may accidentally calculate a maxTimeMS value of 0, which would indicate no timeout
+			// server-side.
+			if remainingTimeout <= rtt90 {
+				return 0, internal.WrapErrorf(ErrDeadlineWouldBeExceeded,
+					"remaining time %v until context deadline is less than or equal to 90th percentile RTT", time.Until(deadline))
+			}
+			return uint64(remainingTimeout/time.Millisecond - rtt90/time.Millisecond), nil
```

Comment:
It's possible for `calculateMaxTimeMS` to return 0 even if `remainingTimeout > rtt90` when the difference is less than 1ms. Consider the following example:
```go
remainingTimeout := 1900 * time.Microsecond
rtt90 := 1800 * time.Microsecond
fmt.Println(remainingTimeout > rtt90) // Prints "true"
fmt.Println(uint64(remainingTimeout/time.Millisecond - rtt90/time.Millisecond)) // Prints "0"
```

Update this to calculate the `maxTimeMS` value first, and then return an error if the value is not positive.

E.g.
```go
maxTime := remainingTimeout - rtt90
// Always round up to the next millisecond value so we never truncate the calculated
// MaxTime value (e.g. 400 microseconds evaluates to 1ms, not 0ms).
maxTimeMS := int64((maxTime + (time.Millisecond - 1)) / time.Millisecond)
if maxTimeMS <= 0 {
	return 0, internal.WrapErrorf(ErrDeadlineWouldBeExceeded, ...)
}
return uint64(maxTimeMS), nil
```

Reply:
Excellent point; added this exact logic in.